java/arcs/core/entity/ForeignReferenceChecker.kt

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+package arcs.core.entity
+
+import arcs.core.data.Schema
+import arcs.core.storage.Reference as StorageReference
+import arcs.core.storage.keys.ForeignStorageKey
+
+/**
+ * Singleton object used to register hooks that can check the validity of external references. A
+ * check is associate with a namespace, represented by an empty Arcs schema. It checks a given id in
+ * this namespace for external validity.
+ */
+object ForeignReferenceChecker {
```

Comment:
Rather than introducing another global singleton, Let's make this a class as well, that's provided to the components that need it.

Reply:
we can discuss more, but that is challenging. Particles need access to this via the foreignReference method below, and we also need it to dereference (which can be done across different hosts) so the proper way would be to use a entity handle + the storage service, but until we have those, I don't know how to avoid a singleton object.

Reply:
Ok I made the change to have ForeignReferenceChecker as a class. EntityHandleManager has a ForeignReferenceChecker which is used when dereferencing.Particles cannot access it, so I removed the check from the foreignReference function (as well as the dereferencer installation). Instead we can do these checks (and install a dereferencer) when we write the entity.
Personally, I think the previous version was simpler to understand and simpler to use (see differences in the tests - you cannot install a derefencer or check validity at reference creation time), as well as less error prone (now you can run into problems if the checker functions are registered in the wrong instance of the ForeignReferenceChecker).
As I said in the doc we can also still get a read-only interface that the singleton implements. And the current delete prop solution uses a dagger set which basically acts like a singleton object. So I don't think that singleton objects are bad per se and I think the first version was cleaner.

I am happy to do this modification just to push things forward, but it comes with some drawbacks as well which will make things more complicated on your side.

Reply:
Thanks for taking a shot at it even though you're not convinced it's the right direction. I saw you set up some time to chat in person, seems like a good idea!